File: server/prompt_tokens.py
"""Stable-id prompt token grammar and compile-time ordinal resolution."""

from __future__ import annotations

import re


PROMPT_TOKEN_KINDS = {
    "subject": ("subjects", "Subject"),
    "picture": ("pictures", "Picture"),
    "video": ("videos", "Video"),
    "audio": ("audios", "Audio"),
}

_TOKEN_RE = re.compile(
    r"@(?P<kind>[a-z][a-z0-9_]{0,63})\("
    r"(?P<source_id>[A-Za-z0-9][A-Za-z0-9._:-]*)\)"
)


def _declarations(raw=None) -> dict:
    if raw is None:
        return {
            kind: {"manifest_key": manifest_key,
                   "label_template": f"<{label} {{n}}>",
                   "physical": kind != "subject"}
            for kind, (manifest_key, label) in PROMPT_TOKEN_KINDS.items()
        }
    return {str(kind): value for kind, value in raw.items()
            if isinstance(kind, str) and isinstance(value, dict)} \
        if isinstance(raw, dict) else {}


def token(kind: str, source_id: str, declarations=None) -> str:
    """Return an authored token for a supported kind and stable source id."""
    normalized_kind = str(kind or "").lower()
    normalized_id = str(source_id or "").strip()
    if normalized_kind not in _declarations(declarations) or not re.fullmatch(
            r"[A-Za-z0-9][A-Za-z0-9._:-]*", normalized_id):
        return ""
    return f"@{normalized_kind}({normalized_id})"


def references(text, declarations=None) -> list[dict]:
    supported = _declarations(declarations)
    return [{"token": match.group(0), "kind": match.group("kind"),
             "source_id": match.group("source_id")}
            for match in _TOKEN_RE.finditer(str(text or ""))
            if match.group("kind") in supported]


def find(text, declarations=None) -> list[str]:
    """Return supported authored token spellings without resolving them."""
    return [value["token"] for value in references(text, declarations)]


# A handle mention written directly in authored prose. Same shape as
# `prompt_context.PROMPT_HANDLE_RE`, which validates a handle at creation, so a
# handle that cannot be written cannot be created either.
#
# Two guards, and both are load-bearing:
#
# * `(?<![A-Za-z0-9_])` — an `@` following a word character is an ADDRESS.
#   `bob@example` must never be read as a mention. This mirrors the boundary
#   test in `writingMentionQuery`, which is where the rule was first settled;
#   anything else may precede an `@`, because prose reaches one after a full
#   stop, a quote, a dash or an opening bracket far more often than after a
#   bare space.
# * `(?![A-Za-z0-9_(])` — the run must END where it ends. The `(` half is what
#   keeps `@subject(id)`, the TOKEN grammar above, out of the handle scanner: a
#   project is free to hold a handle spelled `subject`, and tokens win because
#   they carry a stable id where a handle does not. The word-character half is
#   NOT redundant with the greedy quantifier — without it the engine simply
#   backtracks to a shorter run that clears the `(` guard, and `@subject(id)`
#   was measured matching `subjec`. Both halves, or neither works.
#
# There is deliberately NO dotted qualifier here, unlike the completion probe in
# the browser. A handle in prose is a MENTION and nothing else: mentions are
# authored placements that emit per placement, while every other capability
# keys dedupe on an owner that plain text cannot carry, so `@name.definitions`
# in prose would emit a second definition beside a chip's with no
# `conflicting_emission` to catch it.
_HANDLE_RE = re.compile(r"(?<![A-Za-z0-9_])@([A-Za-z][A-Za-z0-9_]{0,63})(?![A-Za-z0-9_(])")


def handle_mentions(text) -> list[dict]:
    """Every `@handle` run in authored prose, with its span.

    Spans rather than spellings, because the caller substitutes in place and
    needs to know what to replace. Returned in source order.
    """
    return [{"handle": match.group(1), "start": match.start(), "end": match.end()}
            for match in _HANDLE_RE.finditer(str(text or ""))]


def has_handle_mention(text) -> bool:
    """Whether prose carries anything the handle pass would resolve.

    Separate from `handle_mentions` because the compile-entry predicate asks
    only the yes/no question over every channel of every section, and building
    span dicts to throw them away is the wrong shape for that.
    """
    return _HANDLE_RE.search(str(text or "")) is not None


def _unit_label_ordinal(declaration, source_id: str, unit_source_labels) -> int | None:
    """Compatibility fallback for a semantic id with exactly one typed source."""
    template = str((declaration or {}).get("label_template") or "")
    prefix, suffix = template.split("{n}", 1) if "{n}" in template else ("", "")
    ordinals = []
    for value in (unit_source_labels or {}).get(source_id) or []:
        text = str(value)
        if not prefix or not text.startswith(prefix) or not text.endswith(suffix):
            continue
        number = text[len(prefix):len(text) - len(suffix) if suffix else None]
        if number.isdigit():
            ordinals.append(int(number))
    unique = sorted(set(ordinals))
    return unique[0] if len(unique) == 1 else None


def resolve(text, ordinal_manifest, unit_source_labels=None,
            declarations=None) -> tuple[str, list[str]]:
    """Resolve supported tokens without mutating the authored text or manifests.

    Unknown stable ids remain visible in the returned text and are reported in
    first-occurrence order so the compiler can attach blocking diagnostics to
    the owning attachment.
    """
    value = str(text or "")
    manifest = ordinal_manifest if isinstance(ordinal_manifest, dict) else {}
    declared = _declarations(declarations)
    unresolved = []

    def replace(match: re.Match) -> str:
        kind = match.group("kind")
        source_id = match.group("source_id")
        declaration = declared.get(kind)
        if declaration is None:
            return match.group(0)
        manifest_key = str(declaration.get("manifest_key") or "")
        label_template = str(declaration.get("label_template") or "")
        number = (manifest.get(manifest_key) or {}).get(source_id)
        if number is None and declaration.get("physical") is True:
            number = _unit_label_ordinal(declaration, source_id, unit_source_labels)
        try:
            number = int(number)
        except (TypeError, ValueError):
            number = 0
        if number <= 0:
            if source_id not in unresolved:
                unresolved.append(source_id)
            return match.group(0)
        try:
            return label_template.format(n=number)
        except (KeyError, ValueError, IndexError):
            return match.group(0)

    return _TOKEN_RE.sub(replace, value), unresolved

File: web/js/prompt_tokens.js
// Authored prompt tokens store stable ids; provider ordinals are preview-only.

export const PROMPT_TOKEN_KINDS = Object.freeze({
    subject: Object.freeze({ manifestKey: "subjects", labelTemplate: "<Subject {n}>", physical: false }),
    picture: Object.freeze({ manifestKey: "pictures", labelTemplate: "<Picture {n}>", physical: true }),
    video: Object.freeze({ manifestKey: "videos", labelTemplate: "<Video {n}>", physical: true }),
    audio: Object.freeze({ manifestKey: "audios", labelTemplate: "<Audio {n}>", physical: true }),
});

// Deliberate mirrors of `prompt_context.SHOT_ORDINAL_KEY` and
// `prompt_payload.SHOT_LABEL_TEMPLATE`, held in parity by
// tests/test_prompt_tokens.py. Shot is compiler-owned ordering rather than an
// identity or a physical population, so it enters the grammar from its own
// capability declaration on both sides.
const SHOT_ORDINAL_KEY = "shots";
const SHOT_LABEL_TEMPLATE = "[Shot {n}]";

const TOKEN_ID = /^[A-Za-z0-9][A-Za-z0-9._:-]*$/;
const TOKEN_PATTERN = /@([a-z][a-z0-9_]{0,63})\(([A-Za-z0-9][A-Za-z0-9._:-]*)\)/g;

// Mirror of `prompt_tokens._HANDLE_RE`, held in parity by
// tests/test_prompt_tokens.py. The comment there carries the reasoning for
// both guards and for the absence of a dotted qualifier; keep the two literals
// identical rather than "equivalent".
const HANDLE_PATTERN = /(?<![A-Za-z0-9_])@([A-Za-z][A-Za-z0-9_]{0,63})(?![A-Za-z0-9_(])/g;

/** Every `@handle` run in authored prose, with its span, in source order. */
export function promptHandleMentions(text) {
    const value = String(text ?? "");
    const out = [];
    HANDLE_PATTERN.lastIndex = 0;
    for (let match = HANDLE_PATTERN.exec(value); match;
        match = HANDLE_PATTERN.exec(value)) {
        out.push({ handle: match[1], start: match.index,
            end: match.index + match[0].length });
    }
    return out;
}

function tokenDeclarations(raw) {
    if (raw == null) return PROMPT_TOKEN_KINDS;
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) return {};
    return Object.fromEntries(Object.entries(raw).filter(([kind, value]) =>
        typeof kind === "string" && value && typeof value === "object" && !Array.isArray(value)));
}

export function promptTokenDeclarationsFromProfile(profile = {}) {
    const result = {};
    for (const declaration of profile?.identity_kinds || []) {
        const kind = String(declaration?.token_kind || declaration?.key || "").toLowerCase();
        const manifestKey = String(declaration?.ordinal_key || `${declaration?.key || kind}s`);
        const labelTemplate = String(declaration?.referenced_label_template || "");
        if (!kind || !manifestKey || !labelTemplate) continue;
        result[kind] = { manifestKey, labelTemplate, physical: false };
    }
    for (const declaration of profile?.physical_populations || []) {
        const kind = String(declaration?.token_kind || "").toLowerCase();
        const manifestKey = String(declaration?.ordinal_key || "");
        const labelTemplate = String(declaration?.label_template || "");
        if (!kind || !manifestKey || !labelTemplate) continue;
        result[kind] = { manifestKey, labelTemplate, physical: true };
    }
    const shotKind = String(profile?.capabilities?.shot?.token_kind || "").toLowerCase();
    if (shotKind) {
        result[shotKind] = {
            manifestKey: SHOT_ORDINAL_KEY,
            labelTemplate: SHOT_LABEL_TEMPLATE,
            physical: false,
        };
    }
    return result;
}

export function promptToken(kind, sourceId, declarations = null) {
    const normalizedKind = String(kind || "").toLowerCase();
    const normalizedId = String(sourceId || "").trim();
    if (!tokenDeclarations(declarations)[normalizedKind] || !TOKEN_ID.test(normalizedId)) return "";
    return `@${normalizedKind}(${normalizedId})`;
}

function unitLabelOrdinal(declaration, sourceId, unitSourceLabels = {}) {
    const template = String(declaration?.labelTemplate || "");
    const marker = template.indexOf("{n}");
    if (marker < 0) return 0;
    const prefix = template.slice(0, marker);
    const suffix = template.slice(marker + 3);
    const ordinals = [...new Set((unitSourceLabels?.[sourceId] || []).map((value) => {
        const text = String(value);
        if (!prefix || !text.startsWith(prefix) || !text.endsWith(suffix)) return 0;
        const end = suffix ? text.length - suffix.length : text.length;
        const number = text.slice(prefix.length, end);
        return /^\d+$/.test(number) ? Number(number) : 0;
    }).filter((value) => value > 0))].sort((a, b) => a - b);
    return ordinals.length === 1 ? ordinals[0] : 0;
}

export function resolvePromptTokens(text, ordinalManifest = {}, unitSourceLabels = {}, declarations = null) {
    const declared = tokenDeclarations(declarations);
    const unresolvedIds = [];
    const resolvedText = String(text || "").replace(TOKEN_PATTERN, (original, kind, sourceId) => {
        const declaration = declared[kind];
        if (!declaration) return original;
        let ordinal = Number(ordinalManifest?.[declaration.manifestKey]?.[sourceId] || 0);
        if (!ordinal && declaration.physical === true) {
            ordinal = unitLabelOrdinal(declaration, sourceId, unitSourceLabels);
        }
        if (!Number.isInteger(ordinal) || ordinal <= 0) {
            if (!unresolvedIds.includes(sourceId)) unresolvedIds.push(sourceId);
            return original;
        }
        return String(declaration.labelTemplate || "").replace("{n}", String(ordinal));
    });
    return { resolvedText, unresolvedIds };
}
